Attempt to fix split up/down Clojure migrations.

armstnp opened the pull request:
Currently, contrary to the docs, using split up/down Clojure migrations does not run as intended; instead, it runs the *last defined function in the migration file*, which is often going to be `down` by convention.

This is an attempt to guess the intent of a migration by checking for `change` functions explicitly, running it if found, and trying to run the `up` and `down` migrations otherwise. It also fixes a potential issue with `down` migrations being inverted by code intended to invert `change` migrations automatically.

This is an ugly draft, so more would need to be done here to proof it, I expect.

armstnp committed:
Attempt to fix split up/down Clojure migrations.

Currently, contrary to the docs, using split up/down Clojure migrations does not run as intended; instead, it runs the *last defined function in the migration file*, which is often going to be `down` by convention.

This is an attempt to guess the intent of a migration by checking for `change` functions explicitly, running it if found, and trying to run the `up` and `down` migrations otherwise. It also fixes a potential issue with `down` migrations being inverted by code intended to invert `change` migrations automatically.

This is an ugly draft, so more would need to be done here to proof it, I expect.

swlkr commented:
We could do something like this instead:

```clojure
    (let [f (load-file filename-with-path)
          {:syms [down change]} (-> f meta :ns .getName ns-publics)
          ident (fn [] nil)
          contents (apply str (or ((or down ident)) ((or change ident))))]
        {:sql contents
         :filename filename}))))
```

swlkr commented:
This would work for `up` as well if you replace `down` in the destructure and that crazy `or` ident thing

armstnp commented:
That's the first I've seen of `:syms` destructuring; that's perfect! I'll see about getting that expression to work against real migrations, and then make this a formal merge proposal.

armstnp committed:
Clean migration fn selection with destructuring.

If there's a cozier way to wrap that double reset, I'm all ears. Definitely feels like it could be extracted, especially since it behaves like a `with-binding` in this context.

swlkr commented on src/coast/migrations.clj:
```diff
@@ -115,7 +116,15 @@
       (let [f (load-file filename-with-path)
-            output (f)]
+            ;; Figure out what the migration's given us...
+            {:syms [down change] (-> f meta :ns .getName ns-publics)
```
I think this is missing a curly

swlkr commented on src/coast/migrations.clj:
```diff
@@ -115,7 +116,15 @@
+                     down   (do
+                              (reset! coast.db.migrations/rollback? false) ;; don't auto-undo anything,
+                              (down)             ;; because we're gonna run 'down' exactly as provided;
+                              (reset! coast.db.migrations/rollback? true)) ;; and then put it back.
```
Eh the double reset, I say it's fine

armstnp committed:
Fix unmatched brace.

swlkr commented:
Hey, thanks a lot for contributing!

LGTM :shipit:
